### PR TITLE
Explicitly raise an error if field names clashes with types

### PR DESCRIPTION
**Pull request by Viicos**
## Change Summary
Linked: https://github.com/pydantic/pydantic/issues/8240.

I'm raising inside `from_annotated_attribute` so I don't have to duplicate the logic for dataclass, BaseModel etc. But unfortunately I can't provide info to the user regarding the field causing the issue :/

## Checklist

* [ ] The pull request title is a good summary of the changes - it will be used in the changelog
* [ ] Unit tests for the changes exist
* [ ] Tests pass on CI
* [ ] Documentation reflects the changes where applicable
* [ ] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

**Comment by sydney-runkle** [comment]
Do we want to consider a case like second one here as well?

```py
from datetime import date
from typing import Union

from pydantic import BaseModel, Field

class Foo(BaseModel):
    date: date = Field(default_factory=date.today)

f = Foo()
#> RecursionError: maximum recursion depth exceeded while getting the repr of an object

class FooWithUnion(BaseModel):
    date: Union[date, None] = Field(default=None)

f = FooWithUnion()
#> RecursionError: maximum recursion depth exceeded while getting the repr of an object

```

**Comment by Viicos** [comment]
Thanks for the feedback, I'll try to come up with a solution covering all cases. Hopefully it won't involves hacky stuff as this might get tricky

**Comment by Viicos** [comment]
> Do we want to consider a case like second one here as well?

@sydney-runkle where you able to run this code snippet? `Union` is a special form and the arguments are checked to be sure they are a type. I got the following on my end: `TypeError: Union[arg, ...]: each arg must be a type. Got FieldInfo ...`

Note that setting a default value of `None` [will pass](https://github.com/python/cpython/blob/c1c6bedfd3ee66ad208e0cd9cdd732374c95b83d/Lib/typing.py#L136-L137):

```python
from datetime import date
from typing import Union

from pydantic import BaseModel

class FooWithUnion(BaseModel):
    date: Union[date, None] =None

FooWithUnion.model_json_schema()
#> {'properties': {'date': {'default': None, 'title': 'Date', 'type': 'null'}}, 'title': 'B', 'type': 'object'}
# ^ type: null!
``` 
And in this case the issue is silently ignored, so having an explicit error will be beneficial in this case as well

**Comment by sydney-runkle** [comment]
@Viicos,

You're right, I get the following:

```py
from datetime import date
from typing import Union

from pydantic import BaseModel, Field

class FooWithUnion(BaseModel):
    date: Union[date, None] = Field(default=None)

f = FooWithUnion()
#> Union[arg, ...]: each arg must be a type. Got FieldInfo(annotation=NoneType, required=False).
```

Agreed, a more explicit error would be helpful 😄

**Comment by sydney-runkle** [comment]
Makes sense. Curious to hear what @alexmojaki thinks as well.

The case that I mentioned isn't as urgent. As long as we cover the original case and the case that @samuelcolvin mentioned above, that's a pretty significant improvement over what we have now.

**Comment by alexmojaki** [comment]
I'm flattered :smile: 

Here's an idea of how it might be possible to actually fix the problem: https://github.com/pydantic/pydantic/compare/main...alexmojaki:pydantic:field-type-clash

**Comment by Viicos** [comment]
Again, @alexmojaki really clever solution. I've played a bit and this looks promising. I'll continue experimenting and will try to come up with an implementation. Thanks again!

**Comment by Viicos** [comment]
So I've played a bit with it, here is an implementation following @alexmojaki's idea: https://github.com/pydantic/pydantic/compare/main...Viicos:pydantic:allow-name-clashes

Main differences are:
- We need a custom `__setitem__` in case class attributes are accessed in the body:
```python
class Model(BaseModel):
    def func(self):
        pass

    func2 = func  # Python will try to get "func" from our custom namespace, where "func" is still prefixed/mangled
```

- Do not "override" the custom namespace by a plain dict in dynamic model creations.

All tests passes with this implementation, but testing that it can actually use name that clashes is harder as it does not work inside functions (see [here](https://github.com/pydantic/pydantic/issues/7327#issuecomment-1830399262)).

I'll let maintainers decide if they want to support this and continue with this solution, otherwise we can continue on this PR and try to see how we could raise better errors.

**Comment by alexmojaki** [comment]
Does this work with postponed/string annotations? Maybe to support using non-field variables in the class body (which I didn't) you could keep `localns` in `get_cls_type_hints_lenient` but exclude `model_fields`.

> We need a custom `__setitem__` in case class attributes are accessed in the body:

Do you mean `__getitem__`? If so, does this mean the `__getitem__` is called when evaluating normal expressions in the class body, but not when evaluating annotations? That's weird.

> All tests passes with this implementation, but testing that it can actually use name that clashes is harder as it does not work inside functions (see https://github.com/pydantic/pydantic/issues/7327#issuecomment-1830399262).

Is it fair to say that in every case where it doesn't work inside functions, it has never worked, and it also doesn't work with plain classes or standard dataclasses?

You *could maybe* make it work inside functions if you're willing to take the black magic further:

```python
def foo():
  class A:
    pass

  class Meta(type):
    @classmethod
    def __prepare__(metacls, name, bases):
      import inspect
      frame = inspect.currentframe().f_back
      return dict(frame.f_locals)

  class B(metaclass=Meta):
    A: A

  return B


print(foo().__annotations__)
```

Or just use one of those test fixtures that makes a module or something.

**Comment by alexmojaki** [comment]
```python
import inspect
from datetime import date


class D(dict):
    def __getitem__(self, key):
        if not key[:2] == '__' == key[-2:]:
            frame = inspect.currentframe().f_back
            assert frame.f_locals is self
            for namespace in [frame.f_back.f_locals, frame.f_globals, frame.f_builtins]:
                try:
                    value = namespace[key]
                except KeyError:
                    pass
                else:
                    if key in self:
                        print(f'Found {key=} in outer namespace, using that instead of value in class')
                    return value
        return super().__getitem__(key)


class M(type):
    @classmethod
    def __prepare__(metacls, name, bases):
        return D()


def foo():
    class A:
        pass

    class C(metaclass=M):
        class B:
            pass

        A: A = 1
        b: B = 2
        date: date = 3
        int: int = 4
        x: int = 5
        y: int = x

    # print(C.A, C.b, C.date, C.int, C.x, C.y)
    # print(C.__annotations__)
    assert C.A == 1
    assert C.b == 2
    assert C.date == 3
    assert C.int == 4
    assert C.x == 5
    assert C.y == 5
    assert C.__annotations__ == {
        'A': A,
        'b': C.B,
        'date': date,
        'int': int,
        'x': int,
        'y': int,
    }


foo()
```

**Comment by alexmojaki** [comment]
To explain the above code (which I've just updated): we can use the custom namespace to override Python's usual scoping rules and check for a local class variable last (i.e. after enclosing scopes, globals, and builtins) rather than first. Of course that could be weird and surprising in some cases, but we can easily emit warnings/errors when relevant so that the user can see what's happening, and this can be as configurable as we want so that upgrading is still safe. I can explain more but right now I need to go to sleep :laughing:

**Comment by Viicos** [comment]
This looks interesting but I don't know if this level of complexity will be accepted? :/

File: pydantic/fields.py
Comment (sydney-runkle) [reply pydantic/fields.py:326]: @Viicos,

Circling back to this solution - I do think we should go ahead and implement something along the lines of what you have here. Could you use a PydanticUserError here instead, as Samuel mentioned?

Happy to review once you make that change 👍

**Review by sydney-runkle — CHANGES_REQUESTED** [review]

File: pydantic/fields.py
Comment (Viicos) [reply pydantic/fields.py:326]: Yes I think for now this would be the best we can do. It doesn't cover all cases (e.g. [this](https://github.com/pydantic/pydantic/pull/8243#issuecomment-1830485561) and [this](https://github.com/pydantic/pydantic/pull/8243#issuecomment-1830510003)) but this is still better than nothing. Will apply changes soon

Comment (sydney-runkle) [reply pydantic/fields.py:326]: Thanks! Yep, let's stick with this for now, then can consider a more complex solution in the future.

**Comment by sydney-runkle** [comment]
@Viicos,

I know you're busy (helping us with other things, partially 🚀 ), but if you have any time to make a minor change here, we'd love to get this change in with the `2.6.0` release early next week 💯.

**Comment by Viicos** [comment]
> I know you're busy (helping us with other things, partially 🚀 ), but if you have any time to make a minor change here, we'd love to get this change in with the `2.6.0` release early next week 💯.

Will take a look this evening/weekend!

File: tests/test_edge_cases.py
Comment (Viicos) [inline tests/test_edge_cases.py:1386]: As said [here](https://github.com/pydantic/pydantic/pull/8243#issuecomment-1830437786): I just removed the now invalid cases

File: docs/errors/usage_errors.md
Comment (sydney-runkle) [inline docs/errors/usage_errors.md:1024]: Great, thanks for adding these docs 👍

File: docs/errors/usage_errors.md
Comment (sydney-runkle) [inline docs/errors/usage_errors.md:1021]:
```suggestion
## Cannot evaluate type annotation {#unevaluable-type-annotation}
```

**Comment by Viicos** [comment]
Regarding the other mentioned use cases, it could also be implemented (more easily) in a linter plugin, e.g. with flake8 or ruff. There are some interesting rules that could also be applied, e.g. enforcing `field: Annotated[<type>, Field(...)]` instead of `field: <type> = Field(...)`, or any use case where people with large code bases would like to enforce consistency.

File: docs/errors/usage_errors.md
Comment (sydney-runkle) [inline docs/errors/usage_errors.md:1032]:
```suggestion
from datetime import date

from pydantic import BaseModel


class Model(BaseModel):
    date: date = Field(description="A date")
```

File: docs/errors/usage_errors.md
Comment (sydney-runkle) [inline docs/errors/usage_errors.md:1045]:
```suggestion
import datetime

# Or `from datetime import date as _date`

from pydantic import BaseModel


class Model(BaseModel):
    date: datetime.date = Field(description="A date")

```

**Comment by sydney-runkle** [comment]
@Viicos,

Looks like a test is still failing - could you please take a look at that before we merge?

**Comment by sydney-runkle** [comment]
@Viicos,

Hmm, still some linting and testing issues

**Comment by Viicos** [comment]
@samuelcolvin @sydney-runkle 

I created [`flake8-pydantic`](https://github.com/Viicos/flake8-pydantic) yesterday, with a bunch of handy rules. [One of them](https://github.com/Viicos/flake8-pydantic?tab=readme-ov-file#pyd005---field-name-overrides-annotation) hopefully covers most of the use cases we encountered.

Let me know if you have other rules in mind that could be implemented!

**Comment by sydney-runkle** [comment]
@Viicos,

Nice! That's great!

**Pull request merged**
